### DISCUSSION TITLE
em.create() requires properties instead of taking the default value

### DISCUSSION THREAD

**Discussion by alanzhang052** [mikro-orm/mikro-orm#2889] [Q&A]
I have a class entity Post that looks like this:
```
@ObjectType()
@Entity()
export class Post {
    @Field()
    @PrimaryKey()
    id!: number;

    @Field(() => String)
    @Property({ type: "date" })
    createdAt = new Date();

    @Field(() => String)
    @Property({ type: "date", onUpdate: () => new Date() })
    updatedAt = new Date();

    @Field()
    @Property({ type: "text" })
    title!: string;
}
```
When I use em.create(), why does it require the properties `createdAt` and `updatedAt`? I expect the em.create() to just create default values with "new Date()", but that doesn't seem to be the case. Instead, I have to assign values as a parameter in em.create():

```
const post = em.create(Post, {
            createdAt: new Date(),
            updatedAt: new Date(),
            title,
        });
```
Can anyone explain the interaction?

**Comment by B4nan** [comment]
```ts
@Entity()
export class Post {
    [OptionalProps]?: 'createdAt' | 'updatedAt';

    @PrimaryKey()
    id!: number;

    @Property({ type: "date" })
    createdAt = new Date();

    @Property({ type: "date", onUpdate: () => new Date() })
    updatedAt = new Date();

    @Property({ type: "text" })
    title!: string;
}
```

**Reply by alanzhang052** [reply]
Ah, I see, that's how OptionalProps work! I was reading the documentation on it but it didn't give an example so I glossed over the option. I was experimenting with the code, does doing the following also work?

```typescript
@Entity()
export class Post {
    @PrimaryKey()
    id!: number;

    @Property({ type: "date" })
    createdAt ?= new Date();

    @Property({ type: "date", onUpdate: () => new Date() })
    updatedAt ?= new Date();

    @Property({ type: "text" })
    title!: string;
}
```

**Reply by B4nan** [reply]
Yes, you can define things as optional explicitly. The way you write it seems a bit funny, as it looks like some kind of `?=` operator, which it is not :]

**Reply by christopher-caldwell** [reply]
Is there a way to use `OptionalProps` with extending a `BaseEntity`? `createdAt` and `updatedAt` are exactly like this example, but are in the `BaseEntity`. When extending, I am unsure how to do that.

```ts
export abstract class BaseEntity {
  [OptionalProps]?: 'createdAt' | 'updatedAt'

  @Property({ fieldName: 'created_at' })
  createdAt: Date = new Date()

  @Property({ fieldName: 'updated_at', onUpdate: () => new Date() })
  updatedAt: Date = new Date()
}

@Entity()
export class Project extends BaseEntity {
  // TS error here, saying it can only be one of createdAt or updatedAt
  [OptionalProps]?: 'status'

  @Enum(() => Status)
  status: Status = Status.CREATED
}
```

**Reply by christopher-caldwell** [reply]
I assume I cannot use `OptionalProps` with the `BaseEntity`.

**Reply by B4nan** [reply]
Use generics, this is covered in the new getting started (wip) guide already:

https://github.com/mikro-orm/guide#using-entity-constructor

**Reply by christopher-caldwell** [reply]
https://github.com/mikro-orm/guide#generics-to-the-rescue for anyone else that is looking for the specific section.
